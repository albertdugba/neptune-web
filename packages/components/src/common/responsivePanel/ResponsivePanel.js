import PropTypes from 'prop-types';

import { Position, Breakpoint } from '..';
import { useClientWidth } from '../../common/hooks';
import BottomSheet from '../bottomSheet';
import Panel from '../panel';

const ResponsivePanel = ({ anchorRef, arrow, children, className, onClose, open, position }) => {
  const windowReference = typeof window === 'undefined' ? undefined : window;

  const [clientWidth] = useClientWidth({ ref: windowReference });

  if (clientWidth >= Breakpoint.SMALL) {
    return (
      <Panel
        key="panel"
        arrow={arrow}
        open={open}
        position={position}
        anchorRef={anchorRef}
        className={className}
        onClose={onClose}
      >
        {children}
      </Panel>
    );
  }
  return (
    <BottomSheet key="bottomSheet" open={open} className={className} onClose={onClose}>
      {children}
    </BottomSheet>
  );
};

ResponsivePanel.defaultProps = {
  arrow: false,
  className: undefined,
  open: false,
  position: Position.TOP,
};

ResponsivePanel.propTypes = {
  arrow: PropTypes.bool,
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  /** Function attached to document click and keydown. It gets called when a close condition provided is met. */
  onClose: PropTypes.func.isRequired,
  open: PropTypes.bool,
  anchorRef: PropTypes.shape({ current: PropTypes.shape({}) }).isRequired,
  position: PropTypes.oneOf(['top', 'right', 'left', 'bottom']),
};

export default ResponsivePanel;
