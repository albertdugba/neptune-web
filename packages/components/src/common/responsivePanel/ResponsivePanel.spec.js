import { Breakpoint, Position } from '..';
import { render, waitFor } from '../../test-utils';

import ResponsivePanel from './ResponsivePanel';

jest.mock('../bottomSheet', () => {
  const { forwardRef } = require('react');
  return forwardRef(({ open, children }, reference) =>
    open ? (
      <div ref={reference} className="np-bottom-sheet">
        {children}
      </div>
    ) : null,
  );
});

jest.mock('../panel', () => {
  const { forwardRef } = require('react');
  return forwardRef(({ open, children }, reference) =>
    open ? (
      <div ref={reference} className="np-panel np-panel--open">
        {children}
      </div>
    ) : null,
  );
});

describe('ResponsivePanel', () => {
  const props = {
    arrow: true,
    children: 'content',
    open: true,
    position: Position.TOP,
    title: 'title',
    anchorRef: jest.fn(),
    onClose: jest.fn(),
  };
  afterEach(() => {
    jest.clearAllMocks();
  });

  let container;

  describe('on desktop', () => {
    it('renders Panel', async () => {
      await waitFor(() => {
        ({ container } = render(
          <ResponsivePanel {...props} intialOpen>
            children
          </ResponsivePanel>,
        ));
      });

      expect(container).toMatchSnapshot();
    });
  });

  describe('on mobile', () => {
    beforeEach(() => {
      global.innerWidth = Breakpoint.SMALL - 1;
    });

    it('renders bottomSheet', async () => {
      await waitFor(() => {
        ({ container } = render(
          <ResponsivePanel {...props} open>
            children
          </ResponsivePanel>,
        ));
      });

      expect(container).toMatchSnapshot();
    });
  });
});
