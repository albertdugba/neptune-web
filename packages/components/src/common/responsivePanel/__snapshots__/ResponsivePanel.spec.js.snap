// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResponsivePanel on desktop renders Panel 1`] = `
<div>
  <div
    class="np-panel np-panel--open"
  >
    children
  </div>
</div>
`;

exports[`ResponsivePanel on mobile renders bottomSheet 1`] = `
<div>
  <div
    class="np-bottom-sheet"
  >
    children
  </div>
</div>
`;
